src/libOpenImageIO/imageinput.cpp

```diff
@@ -1120,11 +1118,13 @@ std::string
 ImageInput::geterror(bool clear) const
 {
     std::string e;
-    std::string* errptr = m_impl->m_errormessage.get();
-    if (errptr) {
-        e = *errptr;
-        if (clear)
-            errptr->clear();
+    auto iter = input_error_messages.find(this);
+    if (iter != input_error_messages.end()) {
+        e = iter.value();
+        if (clear) {
+            iter.value().clear();
+            iter.value().shrink_to_fit();
```

Comment:
Why shrink_to_fit instead of erase? You're freeing the characters, but not getting rid of the node in the hash table that contains the string itself. You could imagine a pattern of repeated ImageInput constructions (with errors) that would lead to a fuller and fuller hash table containing unusabl empty strings? Versus actually removing the hash table entry when the ImageInput is destroyed by the right thread (which is mostly).

I don't think it's a problem in practice -- to be a real issue, you'd need tons of created and then destroyed ImageInput objects that encountered errors along the way. I'm happy to make that a "cross that very unlikely bridge when we get to it." But I'm curious if you did this because you spotted some big flaw in my proposal that I hadn't considered.

Reply:
Oh I see what you mean now. Yes of course erase makes more sense.